### ISSUE TITLE
how tensorflow deal with the problem of not differentiable ops.

### ISSUE THREAD

**Issue by Lzc6996** [zhongwen/RAM#2]
there are some ops in RAM is not differentiable, and in the paper, use the reward to replace the gradient so that we can do the BP.But in the code, your reward is in the loss, so the reward just like a normal loss and tf will compute the gradient include the ops which is not differentiable. i see that some `tf.stop_gradient` you use. is that mean you just stop the gradient of these ops? if it is true ,i dont know why the model work,because the params in the LocNet is frozen.

**Comment by zhongwen** [comment]
Hi, you may find John Schulman's paper [Gradient Estimation Using Stochastic Computation Graphs](http://papers.nips.cc/paper/5899-gradient-estimation-using-stochastic-computation-graphs.pdf) helpful.

**Comment by Lzc6996** [comment]
thank you for that,and i will read this paper now!

**Comment by Lzc6996** [comment]
Hi,because of your help ,now i know why the loss looks like that. but i still dont understand what the `tf.stop_gradient` do and why should stop. The documentation of `tf.stop_gradient` is not clear to me.if you can tell me , i will appreciate it.

**Comment by zhongwen** [comment]
Hi, glad that you found the paper helpful. The `tf.stop_gradient` makes it work like some constants. the crop_image / glimpse operation is not differentiable w.r.t. the bounding boxes. If you would like to look into some differentiable ones, please see Jaderberg et al. [Spatial Transformer Networks](https://papers.nips.cc/paper/5854-spatial-transformer-networks.pdf) and Johnson et al. [DenseCap: Fully Convolutional Localization Networks for Dense Captioning](http://cs.stanford.edu/people/karpathy/densecap/)

**Comment by Lzc6996** [comment]
i am sorry for that i am still have some question.your stop_gradient is for `mean = tf.stop_gradient(mean)` and `loc = tf.stop_gradient(loc)`. but the get_glimpse is not stop_gradient. if the get_glimpse is not differentiable i think the stop_gradient  should in there,right? 
i dont know if you can speak chinese, and i can describe my question more specificly.

**Comment by zhongwen** [comment]
You may send a personal email to me (zhongwen.s.xu@gmail.com) in Chinese.
